### PR TITLE
Remove gamma from groth16 (depends on #2)

### PR DESCRIPTION
**Pull request by dtebbs**

### CODE REVIEW

**Comment by dtebbs** [comment]
Depends on #2

**Comment by AntoineRondelet** [comment]
The aim of this PR is to modify the shape/structure of the SRS for Groth16 as introduced in https://github.com/zcash/sapling-security-analysis/blob/master/secondmpc.pdf. Doing so would pave the way to an MPC for the SRS generation of Zeth (and other potential projects using Groth16 as a proving system).

File: libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.tcc [hunk 1]
```diff
@@ -396,6 +387,26 @@ r1cs_gg_ppzksnark_keypair<ppT> r1cs_gg_ppzksnark_generator(const r1cs_gg_ppzksna
...
+
+    /* Generate secret randomness */
+    const libff::Fr<ppT> t = libff::Fr<ppT>::random_element();
+    const libff::Fr<ppT> alpha = libff::Fr<ppT>::random_element();
+    const libff::Fr<ppT> beta = libff::Fr<ppT>::random_element();
+    const libff::Fr<ppT> delta = libff::Fr<ppT>::random_element();
+    const libff::G1<ppT> g1_generator = libff::G1<ppT>::one();
+
+    r1cs_gg_ppzksnark_keypair<ppT> key_pair = r1cs_gg_ppzksnark_generator_from_secrets<ppT>(
+        r1cs, t, alpha, beta, delta, g1_generator);
```
Comment (AntoineRondelet) [inline libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.tcc]: Why don't you pass the generator of `g2` as an argument to this function?

File: libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.tcc [hunk 2]
```diff
@@ -313,7 +306,7 @@ r1cs_gg_ppzksnark_keypair<ppT> r1cs_gg_ppzksnark_generator(const r1cs_gg_ppzksna
     libff::leave_block("Generating G1 MSM window table");
 
     libff::enter_block("Generating G2 MSM window table");
-    const libff::G2<ppT> G2_gen = libff::G2<ppT>::random_element();
+    const libff::G2<ppT> G2_gen = libff::G2<ppT>::one();
```
Comment (AntoineRondelet) [inline libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.tcc]: `G2_gen` -> `g2_generator` to keep the naming consistent with the generator of g1.
I think we can pass it as an argument to the function and remove this line

File: libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.tcc (hunk 1 above)
Comment (dtebbs) [reply libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.tcc]: Thanks.  Changed to add `g2_generator` as an argument.

File: libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.hpp [hunk 3]
```diff
@@ -169,38 +169,35 @@ class r1cs_gg_ppzksnark_verification_key {
 public:
```
Comment (AntoineRondelet) [inline libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.hpp]: I'd add https://github.com/zcash/sapling-security-analysis/blob/master/secondmpc.pdf as a reference, and probably a few comments about to explicit the structure of the SRS (VK, PK)

**Comment by AntoineRondelet** [comment]
Quick question @dtebbs 

Should we have:
```
    size_t G1_size() const
    {
        // 1 (alpha_g1) + 1 (beta_g1) + 1 (delta_g1) + A_query.size() + B_query.domain_size() + H_query.size() + L_query.size()
        return 1 + 1 + 1 + A_query.size() + B_query.domain_size() + H_query.size() + L_query.size();
    }

    size_t G2_size() const
    {
        // 1 (delta_g2) + 1 (beta_g2) + B_query.domain_size()
        return 1 + 1 + B_query.domain_size();
    }
```

Here: https://github.com/clearmatics/libsnark/blob/remove-gamma-from-groth16/libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.hpp#L111-L119 ?

**Comment by AntoineRondelet** [comment]
As a side note, I'd add https://github.com/zcash/sapling-security-analysis/blob/master/secondmpc/secondmpc.pdf as a reference here: https://github.com/clearmatics/libsnark/blob/remove-gamma-from-groth16/libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.hpp#L33

In order to get a reference toward the paper that we use to shape the SRS

**Comment by AntoineRondelet** [comment]
This goes outside of the scope of this PR, but I'd also be tempted to refactor the loop: https://github.com/clearmatics/libsnark/blob/remove-gamma-from-groth16/libsnark/zk_proof_systems/ppzksnark/r1cs_gg_ppzksnark/r1cs_gg_ppzksnark.tcc#L280-L284 by:
```
    const size_t Lt_offset = qap.num_inputs() + 1;
    for (size_t i = Lt_offset; i < qap.num_variables(); ++i)
    {
        Lt.emplace_back((beta * At[i] + alpha * Bt[i] + Ct[i]) * delta_inverse);
    }
```

as I don't get the point to change the iteration bounds and shift all indexes by `Lt_offset` in the loop

**Comment by dtebbs** [comment]
> Shouldn't we have:
> ...

@AntoineRondelet Yes, these do seem wrong.  Updated.

**Comment by dtebbs** [comment]
> As a side note, I'd add https://github.com/zcash/sapling-security-analysis/blob/master/secondmpc/secondmpc.pdf as a reference ....

**Comment by dtebbs** [comment]
> ... I'd also be tempted to refactor the loop:

It does seem unecessarily complex.  Changed based on your sample.

**Pull request merged**